### ISSUE TITLE
VAR=value followed by nested function call fails to put VAR into environment

### ISSUE THREAD

**Issue by sckendall2** [att/ast#32]
The following code should print V=1, but doesn't:

```ksh
function f2 { env | grep '^V='; }
function f1 { f2; }
V=1 f1
```

However, `V=1 f2` behaves correctly.

Every ksh93 I tried has the same behavior, including the following:

- Version M-12/28/93e (AIX 6.1 /bin/ksh93)
- Version AJM 93u+ 2012-08-01 (Red Hat Enterprise Linux Server release 7.2 /bin/ksh93)

**Comment by zakukai** [comment]
There's a problem in your example ("env" is an external command, so for it to show $V, $V must be not only set in f2's scope, but also exported)

```
$ function f2 { echo "f2: VF1=$VF1"; }
$ function f1 { f2; }
$ VF1=valkyrie f1
f2: VF1=valkyrie
```

However, local variables are another story:
```
$ function f1 { typeset var VF19; VF19=excalibur; echo "f1: VF1=$VF1, VF19=$VF19"; f2; }
$ function f2 { echo "f2: VF1=$VF1, VF19=$VF19"; }      
$ VF1=valkyrie f1
f2: VF1=valkyrie, VF19=excalibur
f2: VF1=valkyrie, VF19=
```

$VF19 is local to function f1, so it isn't accessible from f2. (Korn Shell implements static scoping: Since $VF19 is not local to f2, $VF19 in f2 always refers to the variable in _global scope_)

```
$ VF19=YF19
gcaswell-vm:/home/onelink/misc_src/shell_bugs $ f1
f1: VF1=, VF19=excalibur
f2: VF1=, VF19=YF19
# f1 has a local binding for VF19 which overrides global scope in f1. But f2 still gets the copy from global scope.
```

However, it seems Korn Shell's "static typing" can be broken by invocation-level bindings:
```
$ unset VF19
$ VF19=YF19 f1      # VF19 is an "invocation-level" binding on f1
f1: VF1=, VF19=excalibur
f2: VF1=, VF19=excalibur
$ echo ${VF19-'{unset}'}
{unset}
```

Basically, when f1 is invoked, it inherits VF19 as an "invocation-level" binding - a temporary global-scope binding (I guess). As a result, f1 can't declare VF19 as local, and when f1 changes the value of VF19, the binding affects f2 as well.

(My tests are on version 93u+ 2012-08-01, RHEL 7)

**Comment by sckendall2** [comment]
@zakukai, your first sentence is incorrect. From the [ksh93 man page](https://linux.die.net/man/1/ksh93): "The environment for any simple-command or function may be augmented by prefixing it with one or more variable assignments." An explicit **export** command is not necessary.

The rest of your comments aren't relevant to this issue.

**Comment by zakukai** [comment]
I guess I misunderstood the issue you were describing. Sorry about that.

**Comment by McDutchie** [comment]
Hmm. The man page is incorrect in any case, whether this behaviour is desired or not:

       The environment for any simple-command or function may be augmented  by
       prefixing it with one or more variable assignments.  A variable assign-
       ment argument is a word of the form identifier=value.  Thus:             
                                                                                
              TERM=450 cmd args                  and                            
              (export TERM; TERM=450; cmd args)                                 

       are equivalent (as far as the  above  execution  of  cmd  is  concerned
       except for special built-in commands listed below - those that are pre-
       ceded with a dagger).                                                    

They are, of course, not equivalent at all if `cmd` is a shell function: if you run a shell function within a subshell, any changes it makes apply to that subshell only, whereas a function run from the main shell can make permanent changes.

In POSIX terms, it's [undefined](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_01) whether (a) variable assignments preceding a function call persist after the function call and (b) whether they are exported or not. So, at least for POSIX functions, the current behaviour is standards-compliant (though undocumented).

**Comment by siteshwar** [comment]
@McDutchie You have a fair point about posix, but current behavior seems inconsistent. I suspect it's a bug in ksh.

I have opened #467 to fix this issue. If any one has objection to this change, I am all ears.

**Comment by sckendall2** [comment]
@siteshwar I don't know the source code, but on the face of it, your fix sounds like it might also fix #38. In any case, thank you!
